### ISSUE TITLE
[REQUEST] Enable optimizations for Rust by default

### ISSUE THREAD

**Issue by kornelski** [mattgodbolt/compiler-explorer#1524]
rust.godbolt.org defaults to no compiler flags, which for `rustc` means completely unoptimized output. This output is incredibly verbose, since even most basic Rust constructs depend on "zero-cost" abstractions, which rely on the optimizer cleaning them up.

**Describe the solution you'd like**
Set `-C opt-level=2` flags by default

**Describe alternatives you've considered**
A checkbox to turn optimizations on/off would work too.

**Additional context**
Most Rust users interact with it via `cargo --release`, so the `rustc` flags are pretty obscure, and hardly anybody knows they need `-C opt-level=2`. I have to google it every time.

**Comment by mattgodbolt** [comment]
Hi: thanks for this.

I strongly disagree that "for C [not optimising] is less of a problem", and even more so for C++ (which is by far the most used language on Compiler Explorer). We don't specify optimisation by default as this makes the code considerably less understandable for those trying to learn assembly. I'd be against making `-C opt-level=2` by default for this reason: and for the more practical reason as existing links to compiler explorer which didn't specify an optimisation level (on purpose) would erroneously pick up the new setting.

I do agree that specifying the options to enable optimisation is trickier in rust than some other languages, but not by much -- `-O` is what I use as it's largely compatible with other compilers' options.  A checkbox could be useful, but that would need to be ported for all compilers and languages to be useful, and would need to encompass things like optimisation levels: at the least levels 1, 2 and 3. Other languages support "optimise for size" too, or "optimise for debug". I'm not sure how best to pick these up, and how not to open the floodgates of having drop-downs/tickboxes for all options and all compilers.

I welcome further thoughts on this: we're always looking to improve the experience on Compiler Explorer.

**Comment by kornelski** [comment]
Thank you for your reply. 

I don't follow the understandability argument. This tiny line:

```rust
for i in 0..10 {}
```

produces this this monstrosity:

https://rust.godbolt.org/z/mjOxsM

This may be made from mostly basic instructions, but the sheer volume makes it really hard to follow. This is a known deficiency of `rustc` — it generates too much code, and relies on LLVM cleaning it up.

Things that I've used explorer for are:

- checking whether a hot loop performs bounds checks or not. It can't be verified in debug mode, which always redundantly checks everything.
- checking whether certain uses of generic code get inlined and optimized out. Even for cases that have specialized implementations in stdlib, so theoretically could be seen without optimizations, the amount of code makes it impractical to follow.
- reporting performance regressions to the Rust team. Again, only relevant in release mode.

Where in C you'd perhaps inspect unoptimized code to see what the compiler was thinking, before the optimizer destroyed it, in Rust MIR output is used instead. For example the same loop as MIR:

```
fn  main() -> () {
    let mut _0: ();                      // return place in scope 0 at src/main.rs:3:11: 3:11
    let mut _1: std::ops::Range<i32>;    // in scope 0 at src/main.rs:4:13: 4:18
    let mut _2: std::ops::Range<i32>;    // in scope 0 at src/main.rs:4:13: 4:18
    let mut _3: std::ops::Range<i32>;    // "iter" in scope 0 at src/main.rs:4:13: 4:18
    let mut _5: std::option::Option<i32>; // in scope 0 at src/main.rs:4:13: 4:18
    let mut _6: &mut std::ops::Range<i32>; // in scope 0 at src/main.rs:4:13: 4:18
    let mut _7: &mut std::ops::Range<i32>; // in scope 0 at src/main.rs:4:13: 4:18
    let mut _8: isize;                   // in scope 0 at src/main.rs:4:8: 4:9
    let mut _10: i32;                    // in scope 0 at src/main.rs:4:8: 4:9
    scope 1 {
        scope 2 {
            let mut _4: i32;             // "__next" in scope 2 at src/main.rs:4:13: 4:18
        }
        scope 3 {
            let _9: i32;                 // "val" in scope 3 at src/main.rs:4:8: 4:9
            scope 4 {
            }
            scope 5 {
                let _11: i32;            // "i" in scope 5 at src/main.rs:4:8: 4:9
            }
            scope 6 {
            }
        }
    }

    bb0: {
        StorageLive(_1);                 // bb0[0]: scope 0 at src/main.rs:4:13: 4:18
        StorageLive(_2);                 // bb0[1]: scope 0 at src/main.rs:4:13: 4:18
        (_2.0: i32) = const 0i32;        // bb0[2]: scope 0 at src/main.rs:4:13: 4:18
                                         // ty::Const
                                         // + ty: i32
                                         // + val: Scalar(Bits { size: 4, bits: 0 })
                                         // mir::Constant
                                         // + span: src/main.rs:4:13: 4:14
                                         // + ty: i32
                                         // + literal: Const { ty: i32, val: Scalar(Bits { size: 4, bits: 0 }) }
        (_2.1: i32) = const 10i32;       // bb0[3]: scope 0 at src/main.rs:4:13: 4:18
                                         // ty::Const
                                         // + ty: i32
                                         // + val: Scalar(Bits { size: 4, bits: 10 })
                                         // mir::Constant
                                         // + span: src/main.rs:4:16: 4:18
                                         // + ty: i32
                                         // + literal: Const { ty: i32, val: Scalar(Bits { size: 4, bits: 10 }) }
        _1 = const std::iter::IntoIterator::into_iter(move _2) -> bb1; // bb0[4]: scope 0 at src/main.rs:4:13: 4:18
                                         // ty::Const
                                         // + ty: fn(std::ops::Range<i32>) -> <std::ops::Range<i32> as std::iter::IntoIterator>::IntoIter {<std::ops::Range<i32> as std::iter::IntoIterator>::into_iter}
                                         // + val: Scalar(Bits { size: 0, bits: 0 })
                                         // mir::Constant
                                         // + span: src/main.rs:4:13: 4:18
                                         // + ty: fn(std::ops::Range<i32>) -> <std::ops::Range<i32> as std::iter::IntoIterator>::IntoIter {<std::ops::Range<i32> as std::iter::IntoIterator>::into_iter}
                                         // + literal: Const { ty: fn(std::ops::Range<i32>) -> <std::ops::Range<i32> as std::iter::IntoIterator>::IntoIter {<std::ops::Range<i32> as std::iter::IntoIterator>::into_iter}, val: Scalar(Bits { size: 0, bits: 0 }) }
    }

    bb1: {
        StorageDead(_2);                 // bb1[0]: scope 0 at src/main.rs:4:17: 4:18
        StorageLive(_3);                 // bb1[1]: scope 0 at src/main.rs:4:13: 4:18
        _3 = move _1;                    // bb1[2]: scope 0 at src/main.rs:4:13: 4:18
        goto -> bb2;                     // bb1[3]: scope 0 at src/main.rs:4:13: 4:18
    }

    bb2: {
        StorageLive(_4);                 // bb2[0]: scope 1 at src/main.rs:4:13: 4:18
        StorageLive(_5);                 // bb2[1]: scope 3 at src/main.rs:4:13: 4:18
        StorageLive(_6);                 // bb2[2]: scope 3 at src/main.rs:4:13: 4:18
        StorageLive(_7);                 // bb2[3]: scope 3 at src/main.rs:4:13: 4:18
        _7 = &mut _3;                    // bb2[4]: scope 3 at src/main.rs:4:13: 4:18
        _6 = _7;                         // bb2[5]: scope 3 at src/main.rs:4:13: 4:18
        _5 = const std::iter::Iterator::next(move _6) -> bb3; // bb2[6]: scope 3 at src/main.rs:4:13: 4:18
                                         // ty::Const
                                         // + ty: for<'r> fn(&'r mut std::ops::Range<i32>) -> std::option::Option<<std::ops::Range<i32> as std::iter::Iterator>::Item> {<std::ops::Range<i32> as std::iter::Iterator>::next}
                                         // + val: Scalar(Bits { size: 0, bits: 0 })
                                         // mir::Constant
                                         // + span: src/main.rs:4:13: 4:18
                                         // + ty: for<'r> fn(&'r mut std::ops::Range<i32>) -> std::option::Option<<std::ops::Range<i32> as std::iter::Iterator>::Item> {<std::ops::Range<i32> as std::iter::Iterator>::next}
                                         // + literal: Const { ty: for<'r> fn(&'r mut std::ops::Range<i32>) -> std::option::Option<<std::ops::Range<i32> as std::iter::Iterator>::Item> {<std::ops::Range<i32> as std::iter::Iterator>::next}, val: Scalar(Bits { size: 0, bits: 0 }) }
    }

    bb3: {
        StorageDead(_6);                 // bb3[0]: scope 3 at src/main.rs:4:17: 4:18
        _8 = discriminant(_5);           // bb3[1]: scope 3 at src/main.rs:4:8: 4:9
        switchInt(move _8) -> [0isize: bb6, 1isize: bb5, otherwise: bb4]; // bb3[2]: scope 3 at src/main.rs:4:8: 4:9
    }

    bb4: {
        unreachable;                     // bb4[0]: scope 0 at src/main.rs:5:2: 5:2
    }

    bb5: {
        StorageLive(_9);                 // bb5[0]: scope 3 at src/main.rs:4:8: 4:9
        _9 = ((_5 as Some).0: i32);      // bb5[1]: scope 3 at src/main.rs:4:8: 4:9
        StorageLive(_10);                // bb5[2]: scope 4 at src/main.rs:4:8: 4:9
        _10 = _9;                        // bb5[3]: scope 4 at src/main.rs:4:8: 4:9
        _4 = move _10;                   // bb5[4]: scope 4 at src/main.rs:4:8: 4:9
        StorageDead(_10);                // bb5[5]: scope 4 at src/main.rs:4:8: 4:9
        StorageDead(_9);                 // bb5[6]: scope 3 at src/main.rs:4:17: 4:18
        StorageDead(_5);                 // bb5[7]: scope 3 at src/main.rs:4:17: 4:18
        StorageDead(_7);                 // bb5[8]: scope 3 at src/main.rs:4:17: 4:18
        StorageLive(_11);                // bb5[9]: scope 3 at src/main.rs:4:8: 4:9
        _11 = _4;                        // bb5[10]: scope 3 at src/main.rs:4:13: 4:18
        StorageDead(_11);                // bb5[11]: scope 3 at src/main.rs:4:20: 4:21
        StorageDead(_4);                 // bb5[12]: scope 1 at src/main.rs:4:20: 4:21
        goto -> bb2;                     // bb5[13]: scope 1 at src/main.rs:4:4: 4:21
    }

    bb6: {
        StorageDead(_9);                 // bb6[0]: scope 3 at src/main.rs:4:17: 4:18
        StorageDead(_5);                 // bb6[1]: scope 3 at src/main.rs:4:17: 4:18
        StorageDead(_7);                 // bb6[2]: scope 3 at src/main.rs:4:17: 4:18
        StorageDead(_4);                 // bb6[3]: scope 1 at src/main.rs:4:20: 4:21
        StorageDead(_3);                 // bb6[4]: scope 0 at src/main.rs:4:17: 4:18
        StorageDead(_1);                 // bb6[5]: scope 0 at src/main.rs:4:17: 4:18
        return;                          // bb6[6]: scope 0 at src/main.rs:5:2: 5:2
    }
}
```

To sum it up: compiler explorer is super useful for Rust, but unoptimized Rust output is not useful.

**Comment by apmorton** [comment]
@mattgodbolt do we have the ability to specify "default options" that will simply be pre-populated into the user options textbox when a compiler is selected?

If we don't already that might be a useful feature and would probably be sufficient for the request here.


Something like:
`if (!userOptions && defaultUserOptions) userOptions = userDefaultOptions;`

**Comment by RabsRincon** [comment]
In the configs, there is a per-language `defaultOptions` property which does this, yes.

**Comment by gilescope** [comment]
It's such a gotcha that `-O` isn't on by default for rustc. It would be really _really_ helpful as if I'm looking at godbolt I'm not really caring about debug mode - I want to focus on performance. And instead of 163 instructions I get 24 instructions which is much clearer as to what's actually going on.

**Comment by kornelski** [comment]
Here's an [example of the confusion](https://www.reddit.com/r/rust/comments/mxhbdq/why_doesnt_clone_on_copy_types_just_copy/). You could say that Rust is *uniquely bad* with its unoptimized output.

If you don't want to change default flags, maybe you could make the UI display a clear indicator that this is unoptimized output and not very relevant to Rust programs?

e.g. if the compiler options field is empty, use "unoptimized_example" as the crate name instead of just "example", so that people will notice why their functions are so awful.

**Comment by mattgodbolt** [comment]
> You could say that Rust is uniquely bad with its unoptimized output

Have you seen C++ output? https://godbolt.org/z/dPqeo6qao we have the same there. It's what the compiler does when you invoke it out of the box without optimization on; and it's what we show. Rust isn't alone in having (by design!) extremely sub-optimal code generation. Inlining is an amazing door-opener to reducing overhead.

Compiler Explorer is designed to show you what the compiler does when you compile your code, along with the options you choose for it. I don't disagree that if we could pre-populate the options box with `-O` that would help; though then you have the reverse problem of "if the user wants to see unoptimized code, if they switch compiler we'll re-populate it with `-O`.

I'd definitely look at a PR that tries to solve this, but just turning on optimization is disingenuous, in my opinion.

**Comment by mattgodbolt** [comment]
> <the loop example>

Alright, that _is_ pretty bad :)  The C++ equivalent is slightly better: https://godbolt.org/z/5Gn7Wrh5f but still nothing compared to the `-O1` version: https://godbolt.org/z/7vj88Tq9j which is just a `ret`.

**Comment by gilescope** [comment]
We could hint more in the "Compiler options..." hint text. Maybe "Compiler options... (curruntly unoptimised)" ?
Also I read somewhere that we could set the default compiler options (so that this only trips each person up only once), but I can't find it in the settings screen?

**Comment by martinxyz** [comment]
I think this problem would be 95% solved with a prominent "compiled without optimization" hint above the output.

This is such a common gotcha, every time I see a Rust question with godbolt link, the answer is "you did not turn on optimization". Latest example: https://www.reddit.com/r/rust/comments/o8qmjc/why_do_stringfromhello_and_helloto_string_produce/

**Comment by partouf** [comment]
The problem with such a hint is that we can't really do this for other languages and compilers. It will have to be a rust exception in the UI. I'm not a fan, but I'm open to it if someone can implement it.

**Comment by camsteffen** [comment]
> Where in C you'd perhaps inspect unoptimized code to see what the compiler was thinking, before the optimizer destroyed it, in Rust MIR output is used instead.

I just came across this trying to figure out how to make Godbolt work for me. If the hint were to include "consider using the MIR viewer" that would be helpful.

**Comment by garbageslam** [comment]
> Most Rust users interact with it via cargo --release, so the rustc flags are pretty obscure, and hardly anybody knows they need -C opt-level=2. I have to google it every time.

Maybe `rustc` should just add some shorter ways to set `-O2` for convenience of users?
Is there a reason, `-O0`, `-O1`, `-O2`, `-O3`, can't just work for rustc?

**Comment by kornelski** [comment]
I've eventually discovered that `-O` works as a shorthand. In hindsight it's logical. It's convenient enough to type. 

However, I still think optimizing by default is a good idea, because Rust users use Cargo instead of `rustc`, and are not expected to know even this option (I didn't!). Cargo uses `--release` instead. Plus as mentioned before, the unoptimized code is an unreadable barf.

**Comment by garbageslam** [comment]
@partouf maybe the right answer is, if the user has not changed the compiler options at all, then can display an arrow or something and text like "^ configure compiler optimizations here", and that arrow goes away if the user makes any changes?

**Comment by WofWca** [comment]
There is "Popular arguments", maybe we can force `-O` (or `-C opt-level=2`) there somehow?

![image](https://user-images.githubusercontent.com/39462442/190634020-1717750f-1245-41c6-b353-1569aaf3ae98.png)

**Comment by workingjubilee** [comment]
I believe `-Copt-level=1`, as a lower optimization level and one that produces output that is still fairly readable, may make sense as a default for Rust. I also believe popping open the Rust compiler with `-Copt-level=0` itself may make a lot of sense. It's subtly attention-getting, even though it quite literally is merely setting it to the default value, and I expect it is by far the most-manipulated compiler flag.

**Comment by samlaf** [comment]
+1 here. Just got bitten by this same issue before googling and finding this.

**Comment by idofilus** [comment]
We can also just add `-O`

> `-O`                  Equivalent to `-C opt-level=2`
`rustc --help`

**Comment by mattgodbolt** [comment]
That may be the case, but CE is about gives default compiler output, and letting the user change that. All other compilers give unoptimised output by default too.

A UI that prepopulates `-O` might be ok, but we don't currently support that very well.
